From 0165a6733f0dbc71f05a6902262a026bbdd1af7a Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Fri, 10 Jun 2022 05:51:15 +0300
Subject: [PATCH] [SW-89971] rom : move tpm flow to spi image desc

As per the updated security specification requirement for gaudi3, tpm
image is moved as a part of regular image descriptor.

The bringup api of ppboot image in rom is re-used for tpm image as well,
which also includes the secondary image processing. Currently it is not
intended that tpm would have a different secondary image, but is kept
for forward compatibility purpose.

Change-Id: I167663cf548b7b7163f3335eab00f875b5c7e156
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 4864f99..b6d614a 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -539,7 +539,8 @@
 {
 	uint32_t act_img_size, image_len, loadaddr_arc, num_crc;
 	unsigned char *image;
-	enum arcboot_rom_progress img_sts_base = PROGRESS_BRINGUP_PCI_PRI;
+	enum arcboot_rom_progress img_sts_base = (img == HL_FW_IMG_TPM_SPI) ?
+			PROGRESS_BRINGUP_TPM_SPI_PRI : PROGRESS_BRINGUP_PCI_PRI;
 	enum dma_error err = DMA_SUCCESS;
 	int ret;
 	enum dma_dest_type dst_type;
@@ -547,7 +548,8 @@
 	arcboot_rom_global.curr_img = cur_img;
 
 	if (cur_img == ARCBOOT_SECONDARY)
-		img_sts_base = PROGRESS_BRINGUP_PCI_SEC;
+		img_sts_base = (img == HL_FW_IMG_TPM_SPI) ?
+				PROGRESS_BRINGUP_TPM_SPI_SEC : PROGRESS_BRINGUP_PCI_SEC;
 
 	arcboot_rom_status_update_bringup(img_sts_base, PROGRESS_BRINGUP_START);
 
@@ -570,7 +572,14 @@
 				desc->loadaddr, dst_type, false);
 	if (err != DMA_SUCCESS) {
 		if (err == DMA_INVALID_CRC_ERROR) {
-			security_report_img_status(cur_img, IMG_STATUS_CRC_FAIL);
+			/*
+			 * TPM validation is after ppboot image validation, hence the
+			 * image status of ppboot can be overwritten by tpm validation logic.
+			 * Hence the PSOC debug registers for image status should not be
+			 * done for TPM image.
+			 */
+			if (img != HL_FW_IMG_TPM_SPI)
+				security_report_img_status(cur_img, IMG_STATUS_CRC_FAIL);
 			arcboot_rom_status_update_bringup(img_sts_base,
 					PROGRESS_BRINGUP_CRC_FAIL);
 		} else if (err == DMA_UNALIGNED_DEST_ADDR) {
@@ -604,7 +613,8 @@
 
 	ret = secureboot_verify_image_hash384(desc->chksum, image, image_len);
 	if (ret) {
-		security_report_img_status(cur_img, IMG_STATUS_SIG_AUTH_FAIL);
+		if (img != HL_FW_IMG_TPM_SPI)
+			security_report_img_status(cur_img, IMG_STATUS_SIG_AUTH_FAIL);
 		return ARCBOOT_BRINGUP_ERROR;
 	}
 
@@ -702,13 +712,37 @@
 	regmap_writel(cold_rst_data.data, mmCOLD_RST_FLOPS_2);
 }
 
+static bool arcboot_rom_tpm_spi(void)
+{
+	uint64_t start_addr;
+	int ret;
+	void (*tpm)();
+
+	if (!arcboot_rom_boot_desc_init_done(HL_FW_BOOT_PTR_SPI))
+	   return false;
+
+	ret = arcboot_rom_bringup_img(HL_FW_IMG_TPM_SPI, &start_addr);
+	if (ret != ARCBOOT_BRINGUP_OK)
+	   return false;
+
+	tpm = (void *)((uint32_t)arcboot_rom_get_arc_addr(start_addr));
+	tpm();
+
+	return true;
+}
+
 void arcboot_rom_handle_tpm(void)
 {
 	if (!security_efuse_tpm_en())
 		return;
 
 	if (!security_efuse_tpm_loc()) {
-		/* TODO: Add erom tpm routine*/
+	   arcboot_rom_status_update(PROGRESS_BRINGUP_TPM_SPI);
+
+	   if (arcboot_rom_tpm_spi())
+		   arcboot_rom_status_update(PROGRESS_ARCBOOT_TPM_SPI_FINISH);
+	   else
+		   arcboot_rom_status_update(PROGRESS_ARCBOOT_TPM_SPI_FAIL);
 	} else {
 		arcboot_rom_status_update(PROGRESS_ARCBOOT_ROM_TPM);
 
diff --git a/include/arcboot.h b/include/arcboot.h
index 4eac596..2b1f47d 100644
--- a/include/arcboot.h
+++ b/include/arcboot.h
@@ -197,84 +197,107 @@
 	PROGRESS_ARCBOOT_ROM_TPM_STARTUP_FAIL = 55,
 	PROGRESS_ARCBOOT_ROM_TPM_PCR_EXTEND_FAIL = 56,
 	PROGRESS_ARCBOOT_ROM_TPM_FINISH = 57,
-	PROGRESS_PPBOOT_START = 58,
-	PROGRESS_PPBOOT_FINISH = 59,
-	PROGRESS_BRINGUP_RAM_PRI = 60,
-	PROGRESS_BRINGUP_RAM_PRI_DMA = 61,
-	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 62,
-	PROGRESS_BRINGUP_RAM_SEC = 63,
-	PROGRESS_BRINGUP_RAM_SEC_DMA = 64,
-	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 65,
-	PROGRESS_ARCBOOT_RAM_START = 66,
-	PROGRESS_ARCBOOT_RECOVERY = 67,
-	PROGRESS_ARCBOOT_RECOVERY_DMA = 68,
-	PROGRESS_ARCBOOT_IN_ROM_RECOVERY = 69,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_START = 70,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_INIT = 71,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_LIMIT_GEN1 = 72,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_IATU_CONFIG = 73,
-	PROGRESS_ARCBOOT_ROM_PCIE_LINK_FINISH = 74,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 75,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 76,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 77,
-	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 78,
-	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 79,
-	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 80,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 81,
-	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 82,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 83,
-	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 84,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 85,
-	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 86,
-	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 87,
-	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 88,
-	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 89,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 90,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 91,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 92,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 93,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 94,
-	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 95,
-	PROGRESS_ARCBOOT_EROM_PCIE = 96,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_START = 97,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_CRC_FAILED = 98,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_UNALIGNED_DEST = 99,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_UNALIGNED_IMG_SIZE = 100,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_ERROR = 101,
-	PROGRESS_ARCBOOT_EROM_PCIE_DMA_FINISH = 102,
-	PROGRESS_ARCBOOT_EROM_PCIE_VERIFICATION_FAIL = 103,
-	PROGRESS_ARCBOOT_EROM_PCIE_START = 104,
-	PROGRESS_ARCBOOT_EROM_PCIE_FINISH = 105,
-	PROGRESS_ARCBOOT_EROM_RECOVERY = 106,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_START = 107,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_CRC_FAILED = 108,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_UNALIGNED_DEST = 109,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_UNALIGNED_IMG_SIZE = 110,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_ERROR = 111,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_FINISH = 112,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_VERIFICATION_FAIL = 113,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_START = 114,
-	PROGRESS_ARCBOOT_EROM_RECOVERY_FINISH = 115,
-	PROGRESS_ARCBOOT_EROM_TPM = 116,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_START = 117,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_CRC_FAILED = 118,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_UNALIGNED_DEST = 119,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_UNALIGNED_IMG_SIZE = 120,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_ERROR = 121,
-	PROGRESS_ARCBOOT_EROM_TPM_DMA_FINISH = 122,
-	PROGRESS_ARCBOOT_EROM_TPM_VERIFICATION_FAIL = 123,
-	PROGRESS_ARCBOOT_EROM_TPM_START = 124,
-	PROGRESS_ARCBOOT_EROM_TPM_FINISH = 125,
-	PROGRESS_ARCBOOT_EROM_IMG0 = 126,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 127,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 128,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 129,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 130,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 131,
-	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 132,
-	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 133,
-	PROGRESS_ARCBOOT_EROM_IMG0_START = 134,
-	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 135,
+	PROGRESS_BRINGUP_TPM_SPI = 58,
+	PROGRESS_BRINGUP_TPM_SPI_PRI = 59,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_TOO_BIG = 60,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_CRC_FAIL = 61,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_DEST_ADDR = 62,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_UNALIGNED_IMG_SIZE = 63,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA_ERROR = 64,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_DMA = 65,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_REVOKED = 66,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_KEY_AUTH_FAIL = 67,
+	PROGRESS_BRINGUP_TPM_SPI_PRI_SECURITY = 68,
+	PROGRESS_BRINGUP_TPM_SPI_SEC = 69,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_TOO_BIG = 70,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_CRC_FAIL = 71,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_DEST_ADDR = 72,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_UNALIGNED_IMG_SIZE = 73,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA_ERROR = 74,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_DMA = 75,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_REVOKED = 76,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_KEY_AUTH_FAIL = 77,
+	PROGRESS_BRINGUP_TPM_SPI_SEC_SECURITY = 78,
+	PROGRESS_ARCBOOT_TPM_SPI_FINISH = 79,
+	PROGRESS_ARCBOOT_TPM_SPI_FAIL = 80,
+	PROGRESS_PPBOOT_START = 81,
+	PROGRESS_PPBOOT_FINISH = 82,
+	PROGRESS_BRINGUP_RAM_PRI = 83,
+	PROGRESS_BRINGUP_RAM_PRI_DMA = 84,
+	PROGRESS_BRINGUP_RAM_PRI_SECURITY = 85,
+	PROGRESS_BRINGUP_RAM_SEC = 86,
+	PROGRESS_BRINGUP_RAM_SEC_DMA = 87,
+	PROGRESS_BRINGUP_RAM_SEC_SECURITY = 88,
+	PROGRESS_ARCBOOT_RAM_START = 89,
+	PROGRESS_ARCBOOT_RECOVERY = 90,
+	PROGRESS_ARCBOOT_RECOVERY_DMA = 91,
+	PROGRESS_ARCBOOT_IN_ROM_RECOVERY = 92,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_START = 93,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_INIT = 94,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_LIMIT_GEN1 = 95,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_IATU_CONFIG = 96,
+	PROGRESS_ARCBOOT_ROM_PCIE_LINK_FINISH = 97,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_START = 98,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_STS_BUSY = 99,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_TIMEDOUT = 100,
+	PROGRESS_ARCBOOT_ROM_D2D_LINK_INIT_SUCCESS = 101,
+	PROGRESS_ARCBOOT_ROM_RECOVERY_END = 102,
+	PROGRESS_ARCBOOT_IN_EROM_RECOVERY = 103,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_START = 104,
+	PROGRESS_INIT_EROM_BOOT_DESC_CRC_FAIL = 105,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_DEST = 106,
+	PROGRESS_INIT_EROM_BOOT_DESC_UNALIGNED_IMG_SIZE = 107,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_ERROR = 108,
+	PROGRESS_INIT_EROM_BOOT_DESC_DMA_FINISH = 109,
+	PROGRESS_INIT_EROM_BOOT_DESC_HDR_MAGIC_FAIL = 110,
+	PROGRESS_INIT_EROM_BOOT_DESC_VERIFICATION_FAIL = 111,
+	PROGRESS_INIT_EROM_BOOT_DESC_FAIL = 112,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_START = 113,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_VERSION_FAIL = 114,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_REVOKED = 115,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_KEY_HASH_FAIL = 116,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_AUTH_FAIL = 117,
+	PROGRESS_EROM_BOOT_DESC_VERIFY_SUCCESS = 118,
+	PROGRESS_ARCBOOT_EROM_PCIE = 119,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_START = 120,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_CRC_FAILED = 121,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_UNALIGNED_DEST = 122,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_UNALIGNED_IMG_SIZE = 123,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_ERROR = 124,
+	PROGRESS_ARCBOOT_EROM_PCIE_DMA_FINISH = 125,
+	PROGRESS_ARCBOOT_EROM_PCIE_VERIFICATION_FAIL = 126,
+	PROGRESS_ARCBOOT_EROM_PCIE_START = 127,
+	PROGRESS_ARCBOOT_EROM_PCIE_FINISH = 128,
+	PROGRESS_ARCBOOT_EROM_RECOVERY = 129,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_START = 130,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_CRC_FAILED = 131,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_UNALIGNED_DEST = 132,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_UNALIGNED_IMG_SIZE = 133,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_ERROR = 134,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_DMA_FINISH = 135,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_VERIFICATION_FAIL = 136,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_START = 137,
+	PROGRESS_ARCBOOT_EROM_RECOVERY_FINISH = 138,
+	PROGRESS_ARCBOOT_EROM_TPM = 139,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_START = 140,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_CRC_FAILED = 141,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_UNALIGNED_DEST = 142,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_UNALIGNED_IMG_SIZE = 143,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_ERROR = 144,
+	PROGRESS_ARCBOOT_EROM_TPM_DMA_FINISH = 145,
+	PROGRESS_ARCBOOT_EROM_TPM_VERIFICATION_FAIL = 146,
+	PROGRESS_ARCBOOT_EROM_TPM_START = 147,
+	PROGRESS_ARCBOOT_EROM_TPM_FINISH = 148,
+	PROGRESS_ARCBOOT_EROM_IMG0 = 149,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_START = 150,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_CRC_FAILED = 151,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_DEST = 152,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_UNALIGNED_IMG_SIZE = 153,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_ERROR = 154,
+	PROGRESS_ARCBOOT_EROM_IMG0_DMA_FINISH = 155,
+	PROGRESS_ARCBOOT_EROM_IMG0_VERIFICATION_FAIL = 156,
+	PROGRESS_ARCBOOT_EROM_IMG0_START = 157,
+	PROGRESS_ARCBOOT_EROM_IMG0_FINISH = 158,
 };
 
 enum security_location {
